### PR TITLE
model: Sort projects by type and then by definition file

### PR DESCRIPTION
**Pull request by sschuberth**
Instead of sorting projects by their full id, only sort by type, and
then by definition file depth / path. This causes root projects to
appear first in the result file, which makes using the id of the first
entry as part of the Jenkins build description (see 0366b45) more
meaningful.

Signed-off-by: Sebastian Schuberth <sebastian.schuberth@bosch.io>

### CODE REVIEW

File: model/src/test/kotlin/ProjectTest.kt [hunk 1]
```diff
@@ -34,6 +34,35 @@ private fun readAnalyzerResult(analyzerResultFilename: String): Project =
...
 
 class ProjectTest : WordSpec({
+    "Projects" should {
+        "be sorted by type" {
+            val npmProject = Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "NPM"))
+            val mavenProject = Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "Maven"))
+
+            val projects = listOf(npmProject, mavenProject)
+
+            projects.sorted() shouldBe listOf(mavenProject, npmProject)
```
Comment (fviernau) [inline model/src/test/kotlin/ProjectTest.kt:44]: Consider using `shouldContainExactly()` to make the test case not require the type `List` (without need)?
(Same below)

File: model/src/test/kotlin/ProjectTest.kt [hunk 2]
```diff
@@ -34,6 +34,35 @@ private fun readAnalyzerResult(analyzerResultFilename: String): Project =
...
+                id = Identifier("Maven", "GroupA", "Artifact", "1.0.0"),
+                definitionFilePath = "nested/pom.xml"
+            )
+
+            val rootProject = Project.EMPTY.copy(
+                id = Identifier("Maven", "GroupB", "Artifact", "1.0.0"),
+                definitionFilePath = "pom.xml"
+            )
+
+            val projects = listOf(nestedProject, rootProject)
```
Comment (fviernau) [inline model/src/test/kotlin/ProjectTest.kt:60]: Consider turning `projects` into `sortedProjects` to get AAA sections / blocks.

File: model/src/main/kotlin/Project.kt [hunk 3]
```diff
@@ -155,7 +155,25 @@ data class Project(
     /**
      * A comparison function to sort projects by their identifier.
      */
-    override fun compareTo(other: Project) = id.compareTo(other.id)
+    override fun compareTo(other: Project): Int {
+        // First sort by package type.
+        var sign = id.type.compareTo(other.id.type)
+
+        if (sign == 0) {
```
Comment (fviernau) [inline model/src/main/kotlin/Project.kt:162]: Consider returning early here to eliminate some nesting.

File: model/src/main/kotlin/Project.kt [hunk 4]
```diff
@@ -155,7 +155,25 @@ data class Project(
     /**
      * A comparison function to sort projects by their identifier.
      */
-    override fun compareTo(other: Project) = id.compareTo(other.id)
+    override fun compareTo(other: Project): Int {
```
Comment (fviernau) [inline model/src/main/kotlin/Project.kt:158]: Would it make sense to update the kDoc of the function to mention the definition file path?

Comment (fviernau) [inline model/src/main/kotlin/Project.kt:158]: Are you aware of `compareToBuilder` like `org.apache.commons.lang3.builder.CompareToBuilder` or rather did you consider using one?

Comment (sschuberth) [reply model/src/main/kotlin/Project.kt:158]: I was only aware of Guava's *ComparisonChain*, but didn't want to introduce a (big) dependency for only this one (small) use anyway.

Comment (sschuberth) [reply model/src/main/kotlin/Project.kt:158]: Absolutely.

File: model/src/main/kotlin/Project.kt (hunk 3 above)
Comment (sschuberth) [reply model/src/main/kotlin/Project.kt:162]: Done.

File: model/src/test/kotlin/ProjectTest.kt (hunk 1 above)
Comment (sschuberth) [reply model/src/test/kotlin/ProjectTest.kt:44]: Done.

File: model/src/test/kotlin/ProjectTest.kt (hunk 2 above)
Comment (sschuberth) [reply model/src/test/kotlin/ProjectTest.kt:60]: `projects` actually is the *un*sorted list of projects. Do you mean to put the the *actual* or the *expected* list into "sortedProjects"?

Comment (sschuberth) [reply model/src/test/kotlin/ProjectTest.kt:62]: I've gone with the expected sorter projects now.

Comment (fviernau) [reply model/src/test/kotlin/ProjectTest.kt:62]: What I had in mind was something like the following two.
I actually liked your previous change a bit more but I'm ok also as-is.

```
val npmProject = Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "NPM"))
val mavenProject = Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "Maven"))

val sortedProjects = listOf(npmProject, mavenProject).sorted()

sortedProjects shouldContainExactly(mavenProject, npmProject)
```

```
val projects = listOf(
  Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "NPM")),
  Project.EMPTY.copy(id = Identifier.EMPTY.copy(type = "Maven"))
)

val sortedProject = projects.sorted()

sortedProjects shouldContainExactly(mavenProject, npmProject)
```

Comment (sschuberth) [reply model/src/test/kotlin/ProjectTest.kt]: Done.

File: model/src/main/kotlin/Project.kt (hunk 4 above)
Comment (mnonnenmacher) [reply model/src/main/kotlin/Project.kt:159]: Did you consider using `compareBy` instead?
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.comparisons/compare-by.html

Kotlin also provides a few more helpers for writing comparators that might be handy here:
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-comparator/

**Comment by sschuberth** [comment]
> Can you add an example how that change would change one of the expected results files?

I was planning to wait for failing tests for that purpose (and thus converted the PR to a draft again).

**Comment by sschuberth** [comment]
> What other places would be affected by changing the comparator, in addition to the list of projects in the analyzer result?

As the new version of this PR shows, reports might also be affected.

File: model/src/main/kotlin/Project.kt [hunk 5]
```diff
@@ -94,6 +94,11 @@ data class Project(
             homepageUrl = "",
             scopes = sortedSetOf()
         )
+
+        private val PATH_STRING_COMPARATOR = compareBy<String>({ path -> path.count { it == '/' } }, { it })
+        private val COMPARARTOR = compareBy<Project> { it.id.type }
```
Comment (mnonnenmacher) [inline model/src/main/kotlin/Project.kt:99]: "COMPARA~~R~~TOR"

**Comment by mnonnenmacher** [comment]
I'm not so happy with how this change affects the static HTML report. Before the identifiers for the projects at the bottom were at least alphabetically sorted in the index of the report:

```
Gradle:org.ossreviewtoolkit:a1:1.0
Gradle:org.ossreviewtoolkit:a2:1.0
Gradle:org.ossreviewtoolkit:a:1.0
Gradle:org.ossreviewtoolkit:b1:1.0
Gradle:org.ossreviewtoolkit:b2:1.0
Gradle:org.ossreviewtoolkit:b:1.0
Gradle:org.ossreviewtoolkit:c:1.0
Gradle:org.ossreviewtoolkit:d:1.0
Gradle:org.ossreviewtoolkit:root:1.0
```

This is not ideal, because it hides the hierarchy of the projects, but at least it has an easy to understand order.

Now the order seems to be random (in a real project they would likely not have hierarchical names like I used in this example, so it would seem even more random):

```
Gradle:org.ossreviewtoolkit:root:1.0
Gradle:org.ossreviewtoolkit:a:1.0
Gradle:org.ossreviewtoolkit:b:1.0
Gradle:org.ossreviewtoolkit:c:1.0
Gradle:org.ossreviewtoolkit:d:1.0
Gradle:org.ossreviewtoolkit:a1:1.0
Gradle:org.ossreviewtoolkit:a2:1.0
Gradle:org.ossreviewtoolkit:b1:1.0
Gradle:org.ossreviewtoolkit:b2:1.0
```

If we change sorting of projects to look at the hierarchy, it should show nested projects next to each other, like this:

```
Gradle:org.ossreviewtoolkit:root:1.0
Gradle:org.ossreviewtoolkit:a:1.0
Gradle:org.ossreviewtoolkit:a1:1.0
Gradle:org.ossreviewtoolkit:a2:1.0
Gradle:org.ossreviewtoolkit:b:1.0
Gradle:org.ossreviewtoolkit:b1:1.0
Gradle:org.ossreviewtoolkit:b2:1.0
Gradle:org.ossreviewtoolkit:c:1.0
Gradle:org.ossreviewtoolkit:d:1.0
```

An the hierarchy should also be visualized by adding indentation.

I'm also not so happy with the approach of changing the project sorting for the whole of ORT only for convenience in the Jenkinsfile, I think if we change the sorting it should be in a way that makes sense when listing the projects and not only look at what we want to be first in the list.

What do you think of those alternatives?
* Implement finding the root project in the Jenkinsfile, finding the project with the least amount of "/" in the definition file path should be easy.
* Only apply the new sorting to the ORT result file, but not globally.

```
    val projects = sortedSetOf(
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:root:1.0"),
            definitionFilePath = "build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:a:1.0"),
            definitionFilePath = "a/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:a1:1.0"),
            definitionFilePath = "a/1/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:a2:1.0"),
            definitionFilePath = "a/2/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:b:1.0"),
            definitionFilePath = "b/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:b1:1.0"),
            definitionFilePath = "b/1/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:b2:1.0"),
            definitionFilePath = "b/2/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:c:1.0"),
            definitionFilePath = "c/build.gradle.kts"
        ),
        Project.EMPTY.copy(
            id = Identifier("Gradle:org.ossreviewtoolkit:d:1.0"),
            definitionFilePath = "d/build.gradle.kts"
        )
    )
```

**Comment by sschuberth** [comment]
>     * Implement finding the root project in the Jenkinsfile, finding the project with the least amount of "/" in the definition file path should be easy.

I went for that approach: https://github.com/oss-review-toolkit/ort/pull/2583

**Pull request closed without merging**